acme service: letsencrypt cert auto-renewal with simp_le

fpletz opened the pull request:
This new service invokes `simp_le` for a defined set of certs on a regular basis with a systemd timer. `simp_le` is smart enough to handle account registration, domain validation and renewal on its own. The only thing required is an existing HTTP server that serves the path `/.well-known/acme-challenge` from the webroot cert parameter.
### Example

```
  security.acme.certs."foo.example.com" = {
    webroot = "/var/www/challenges";
    extraDomains = { "www.example.com" = null; };
    email = "foo@example.com";
    validMin = 2592000;
    renewInterval = "weekly";
  };
```
### Example Nginx vhost

```
  services.nginx.appendConfig = ''
    http {
      server {
        server_name _;
        listen 80;
        listen [::]:80;

        location /.well-known/acme-challenge {
          root /var/www/challenges;
        }

        location / {
          return 301 https://$host$request_uri;
        }
      }
    }
  '';
```
### Open TODOs
- [x] Notify/reload/restart dependent services
- [x] ~~Provide a basic HTTP Server for verification if host doesn't provide one (e.g. python SimpleHTTPServer)~~
- [x] Don't run `simp_le` as root and use cert-specific owner/group/mode at least for private keys

Feedback and suggestions for improvements would be much appreciated.

fpletz committed:
simp_le service: letsencrypt cert auto-renewal

This new service invokes `simp_le` for a defined set of certs on a regular
basis with a systemd timer. `simp_le` is smart enough to handle account
registration, domain validation and renewal on its own. The only thing
required is an existing HTTP server that serves the path
`/.well-known/acme-challenge` from the webroot cert parameter.

Example:

  services.simp_le.certs."foo.example.com" = {
    webroot = "/var/www/challenges";
    extraDomains = [ "www.example.com" ];
    email = "foo@example.com";
    validMin = 2592000;
    renewInterval = "weekly";
  };

Example Nginx vhost:

  services.nginx.appendConfig = ''
    http {
      server {
        server_name _;
        listen 80;
        listen [::]:80;

        location /.well-known/acme-challenge {
          root /var/www/challenges;
        }

        location / {
          return 301 https://$host$request_uri;
        }
      }
    }
  '';

lethalman commented on nixos/modules/services/security/simp_le.nix:
```diff
@@ -0,0 +1,133 @@
+          mkdir -p "${cfg.directory}/${cert}"
+        '';
+        script = ''
+          WEBROOT="${optionalString (data.webroot == null) data.webroot}"
```
`WEBROOT` will be always empty with that null check...

abbradar commented:
Do we really want separate `renewInterval`s and `validMin`s for `certOpts`? I can imagine no use case for this.

abbradar committed:
simp_le: 20151205 -> 20151207

abbradar committed:
nixos/simp_le: improve configuration options

abbradar committed:
nixos/simp_le: use /var/lib/simp_le as root dir by default

/etc on NixOS is regenerated on boot and there was movement
towards making it read-only -- so let's keep dynamic state elsewhere.

abbradar commented:
I have several patches on top of your branch at https://github.com/abbradar/nixpkgs/commits/simp_le-new -- what do you think?

globin commented:
@abbradar I'd always want to renew a certain time before the certificate expires.

abbradar commented:
Sorry, bad wording. What I meant is: do we want those settings to be different for every domain?

edolstra commented:
The option name `services.simp_le.certs` seems ugly to me. Isn't the use of `simp_le` just an implementation detail? Might be nicer to call it something like `security.letsEncrypt.certs`.

jgillich commented:
@edolstra Let's Encrypt is also just a [ACME](https://en.wikipedia.org/wiki/Automated_Certificate_Management_Environment) provider, so we should probably just call it `acme`. Would surely avoid confusion down the road when (if) other CAs support the protocol.

jeaye commented:
+1 Ah, I'd love to see this merged in.

edolstra commented:
@jgillich Yeah, sounds good.

domenkozar commented:
Good job :+1: Let's fix small things based on feedback and merge this.

fpletz commented:
@abbradar Sorry for not getting back to you sooner, will look at this later today and incorporate your changes. Thanks! :beers:

abbradar commented:
One of remaining things that bothers me is my own usage of `sudo` which is affected by `/etc/sudoers`. Is this a bad thing (e.g., can someone make such a config (perhaps invalid one) that `sudo` will reject dropping `root` privileges)? Would `su` be better?

fpletz committed:
nixos/simp_le: Use systemd for setting user and group

This is much cleaner and we don't depend on sudo.

fpletz committed:
nixos/simp_le: Rename to security.acme

fpletz commented:
@abbradar Thanks for your patches. I've rebased and updated this PR with those.

Other changes:
 - `sudo` was removed and systemd will set the user and the group instead. A little hackery was needed to run the `postRun` hook as root in the `postStop` systemd hook only if the exit code is 0.
 - `services.simp_le` was renamed to `security.acme` as suggested by @edolstra & @jgillich.

abbradar commented:
@fpletz I quite like your hack, nice!

One little thing (see my comment above): do we want `renewIntervals` and `validMins` to be different for every domain? I can't imagine any use case for this, and to simplify changing them for user we may want to move them to the top-level `security.acme`.

domenkozar commented on nixos/modules/security/acme.nix:
```diff
@@ -0,0 +1,193 @@
+
+      validMin = mkOption {
+        type = types.int;
+        default = 2592000;
```
for easier understanding I'd do `30 * 24 * 3600`

domenkozar commented:
Would be extremely nice to add `meta.doc` attribute to the module with a minimal example for nginx.

fpletz committed:
nixos/acme: Improve documentation

fpletz committed:
nixos/acme: validMin & renewInterval aren't cert-specific

fpletz committed:
nixos/acme: Add module documentation

fpletz commented:
@abbradar You're right about `renewInterval` and `validMin`. Fixed.

@domenkozar Excellent idea. Added. :smile:

abbradar commented:
Thanks! No more complaints from me, I think this is ready. BTW I haven't thought of  (or seen) adding `meta.maintainers` to a NixOS module this way -- very nice, I'll steal this idea ~_^.

domenkozar commented:
Awesome work. As soon as someone confirms it works I'm +1 for merging (sorry, a bit busy now but I'll use this asap)

abbradar commented:
Pushed with a small fix in https://github.com/NixOS/nixpkgs/commit/52505823969c9f7462fbb7efd5f36e32ca92b581. Works for me both for initial certificate and for checking (when `postRun` hook is skipped). Thank you for the work, @fpletz!

jeaye commented:
Great work, folks! Stoked to see this one come in.

fpletz commented:
Thanks for all the feedback! We have some more stuff coming up to improve the nginx module and optional automatic ACME support. :)

jeaye commented:
On Sun, Dec 13, 2015 at 06:42:53AM -0800, Franz Pletz wrote:
>    Thanks for all the feedback! We have some more stuff coming up to improve
>    the nginx module and optional automatic ACME support. :)

Why's nginx getting all the love? What about the apache users? :D

fpletz commented:
Well, if you look at the apache module and compare it to nginx, apache got all the love in the past. :) But we will probably hook in Apache and the other supported HTTP servers, too, if their respective modules already have support for vhosts.

jeaye commented:
Aside from being on unstable, what's needed in order to use this now? A simple config lands me with

```text
error: The option `security.acme' defined in `/etc/nixos/service/acme.nix' does not exist.
(use ‘--show-trace’ to show detailed location information)
```

Config:

```nix
  security.acme =
  {
    certs =
    {
      "pastespace.org" =
      {
        webroot = "/home/http/acme.pastespace.org";
        extraDomains =
        {
          "www.pastespace.org" = null;
          "mail.pastespace.org" = null;
        };
        email = "contact@jeaye.com";
        user = "acme";
        group = "acme";
        postRun = "systemctl reload httpd.service";
      };
    };
  };
```

jgillich commented:
It is in master, I dont think it's in unstable already.


On Mon, Dec 14, 2015, at 08:55 AM, jeaye wrote:
> Aside from being on unstable, what's needed in order to use this now?
> A simple config lands me with


> error: The option `security.acme' defined in
> `/etc/nixos/service/acme.nix' does not exist.
(use ‘--show-trace’ to show detailed location information)
>
> Config:


> security.acme =
  {  certs =  {    "pastespace.org" =    {      webroot =
  "/home/http/acme.pastespace.org";      extraDomains =      {
  "www.pastespace.org" = null;        "mail.pastespace.org" = null;
  };      email = "contact@jeaye.com";      user = "acme";      group =
  "acme";      postRun = "systemctl reload httpd.service";    };  }; };
> — Reply to this email directly or view it on GitHub[1].


>


Links:

  1. https://github.com/NixOS/nixpkgs/pull/11506#issuecomment-164371664
